tensorflow/core/graph/mkl_layout_pass.cc

```diff
@@ -2391,6 +2646,150 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+      // Fill inputs other than "x":
+      nb.Input(mklop_in[i].first, mklop_in[i].second);
+    }
+  }
+
+  copy_attrs(const_cast<const Node*>(mklop), &nb, true);
+  nb.Attr("data_format", data_format);
+
+  // Copy the device assigned to old node to new node.
+  nb.Device(mklop->def().device());
+
+  // Create node.
+  Node* new_node;
+  TF_CHECK_OK(nb.Finalize(&**g, &new_node));
+  CHECK_NOTNULL(new_node);
+
+  // Fill outputs.
+  for (const Edge* e : transpose_to_nchw->out_edges()) {
+    if (!e->IsControlEdge()) {
+      const int kConv2DWithBiasOutputSlot = 0;
+      CHECK_NOTNULL((*g)->AddEdge(new_node, kConv2DWithBiasOutputSlot, e->dst(),
+                                  e->dst_input()));
+    }
+  }
+
+  // Copy device assigned to old node to new node.
+  new_node->set_assigned_device_name(mklop->assigned_device_name());
+
+  (*g)->RemoveNode(transpose_to_nhwc);
+  (*g)->RemoveNode(mklop);
+  (*g)->RemoveNode(transpose_to_nchw);
+
+  return Status::OK();
+}
+
+Status MklLayoutRewritePass::FuseNode(
+    std::unique_ptr<Graph>* g, std::vector<Node*>& nodes,
+    const MklLayoutRewritePass::FusionInfo fi) {
+  return fi.fuse_func(g, nodes, fi.copy_attrs);
+}
+
+std::tuple<bool, std::vector<Node*>, const MklLayoutRewritePass::FusionInfo>
+MklLayoutRewritePass::CheckForNodeFusion(Node* a) const {
+  bool found_pattern = false;
+  std::vector<Node*> nodes;
+  const FusionInfo* fi_ptr = nullptr;
+
+  for (auto fi = finfo_.begin(); fi != finfo_.end(); ++fi) {
+    assert(fi->ops.size() == fi->node_checkers.size());
+    nodes.clear();
+    fi_ptr = &*fi;
+    //
+    // Make sure node "a" and its succeding nodes (b, c ...), match the pattern
+    // defined in fusion info (ops[0], ops[1], ...),
+    // aka. "a->b->c" matches "op1->op2->op3"
+    //
+
+    // Initialize "current_node" as node "a".
+    Node* current_node = a;
+    for (auto node_index = 0; node_index < fi->node_checkers.size();
+         ++node_index) {
+      // Make sure current node meet the requirement of corresponding node
+      // checker.
+      auto check_node = fi->node_checkers[node_index];
+      if (current_node == nullptr ||
+          (check_node && check_node(current_node) == false)) {
+        found_pattern = false;
+        nodes.clear();
+        break;
+      }
+
+      // Add current_node to "fusion_nodes":
+      nodes.push_back(current_node);
+
+      // If current node is not the last node we want to check, check next node.
+      if (node_index != fi->node_checkers.size() - 1) {
+        // Find current node's direct descendant, which will be used in next
+        // iteration.
+        auto check_next_node = fi->node_checkers[node_index + 1];
+        for (const Edge* e : current_node->out_edges()) {
+          if (!e->IsControlEdge()) {
+            Node* candidate_node = e->dst();
+
+            if (check_next_node(candidate_node) == false) {
+              current_node = nullptr;
+            } else {
+              current_node = candidate_node;
```

Comment:
Greedy matching like this could miss a pattern. For example, we are looking for a pattern X->Y->Z. Found a node X which has two output edges to two nodes Y1 and Y2. Y1 doesn't connect to Z. Y2 connects to Z. This loop would add Y1 into the array `nodes`, ignore Y2, and will eventually miss a match X->Y2->Z.

This issue won't happen with TransposeMklOpTranspose because the fusion rules only fuse when all the nodes only have one output data edge. I'm asking for a future/more generic case.